src/ratelimit/RateLimitController.ts

```diff
@@ -0,0 +1,197 @@
+  RateLimitMessenger<RateLimitedApis>
+> {
+  private implementations;
+
+  private rateLimitTimeout;
+
+  private rateLimitCount;
+
+  /**
+   * Creates a RateLimitController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.implementations - Mapping from API type to API implementation.
+   * @param options.rateLimitTimeout - The time window in which the rate limit is applied.
+   * @param options.rateLimitCount - The amount of notifications an origin can show in the rate limit time window.
+   */
+  constructor({
+    // @todo Pick some sane defaults for this
+    rateLimitTimeout = 5000,
+    rateLimitCount = 3,
+    messenger,
+    state,
+    implementations,
+  }: {
+    rateLimitTimeout?: number;
+    rateLimitCount?: number;
+    messenger: RateLimitMessenger<RateLimitedApis>;
+    state?: Partial<RateLimitState<RateLimitedApis>>;
+    implementations: RateLimitedApis;
+  }) {
+    const defaultState = {
+      requests: Object.keys(implementations).reduce(
+        (acc, key) => ({ ...acc, [key]: {} }),
+        {} as Record<keyof RateLimitedApis, Record<string, number>>,
+      ),
+    };
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.implementations = implementations;
+    this.rateLimitTimeout = rateLimitTimeout;
+    this.rateLimitCount = rateLimitCount;
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:call` as const,
+      (
+        origin: string,
+        type: keyof RateLimitedApis,
+        ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]>
+      ) => this.call(origin, type, ...args),
+    );
+  }
+
+  /**
+   * Calls an API if the requesting origin is not rate-limited.
+   *
+   * @param origin - The requesting origin.
+   * @param type - The type of API call to make.
+   * @param args - Arguments for the API call.
+   * @returns `false` if rate-limited, and `true` otherwise.
+   */
+  async call(
+    origin: string,
+    type: keyof RateLimitedApis,
+    ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]>
+  ): Promise<RateLimitWrapper> {
+    if (this._isRateLimited(type, origin)) {
+      return { isRateLimited: true };
+    }
+    this._recordRequest(type, origin);
+
+    const implementation = this.implementations[type];
+
+    if (!implementation) {
+      throw new Error('Invalid api type');
+    }
+
+    const result = await implementation(...args);
+
+    return { isRateLimited: false, result };
+  }
+
+  /**
+   * Checks whether an origin is rate limited for the a specific API.
+   *
+   * @param api - The API the origin is trying to access.
+   * @param origin - The origin trying to access the API.
+   * @returns `true` if rate-limited, and `false` otherwise.
+   */
+  _isRateLimited(api: keyof RateLimitedApis, origin: string) {
```

Comment:
Are these intended to be public, sorta-public or private? If they're intended to be private, what do you think about using the `private` keyword to mark them so?

Reply:
Sure thing! I wasn't sure what sort of pattern to use for that as it seems to differ a bit from controller to controller 👍

Reply:
Done!